### PR TITLE
Add Jetson cuda variants for arm

### LINKED ISSUE #2408: Add binary support for Nvidia Jetson Orin - JetPack 6
I believe Ollama is a great project, I have tried different ideas to try get Ollama to utilise the GPU, but still uses CPU.
I have currently flashed Jetpack 6 DP onto the AGX ORIN Dev Kit. I believe this jetpack version will help Ollama use the GPU easier, if you are able to add support for it.

```shell
nvcc --version
``` 
```shell
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Tue_Aug_15_22:08:11_PDT_2023
Cuda compilation tools, release 12.2, V12.2.140
Build cuda_12.2.r12.2/compiler.33191640_0
``` 

```shell
nvidia-smi
``` 
```shell
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 540.2.0                Driver Version: N/A          CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Orin (nvgpu)                  N/A  | N/A              N/A |                  N/A |
| N/A   N/A  N/A               N/A /  N/A | Not Supported        |     N/A          N/A |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+
``` 

Thank you

### PR DESCRIPTION
**Pull request by dhiltgen**
This adds new variants for arm64 specific to Jetson platforms

Fixes #2408 #4693 #5100 #4861

Updated to layer on #5631 to reduce the payload size so we're not at risk of exceeding the limits.

After extracting the tgz:
```
% ls -F
cuda/  cuda_jetpack5/  cuda_jetpack6/  ollama*  ollama-linux-arm64.tgz
% ls -lh ollama-linux-arm64.tgz
-rw-r--r-- 1 daniel daniel 1.2G Jul 12 09:09 ollama-linux-arm64.tgz
% ls -lh ollama
-rwxr-xr-x 1 daniel daniel 245M Jul 12 08:58 ollama
% du -sh cuda*
356M	cuda
519M	cuda_jetpack5
552M	cuda_jetpack6
```

`"Dynamic LLM libraries [cuda_jetpack6 cuda_v11 cpu cuda_jetpack5]"`

```
time=2024-07-12T09:16:54.654-07:00 level=INFO source=server.go:383 msg="starting llama server" cmd="/tmp/ollama2043741824/runners/cuda_jetpack5/ollama_llama_server --model /home/daniel/.ollama/models/blobs/sha256-6a0746a1ec1aef3e7ec53868f220ff6e389f6f8ef87a01d77c96807de94ca2aa --ctx-size 8192 --batch-size 512 --embedding --log-disable --n-gpu-layers 33 --verbose --parallel 4 --port 45659"
time=2024-07-12T09:16:54.654-07:00 level=DEBUG source=server.go:398 msg=subprocess environment="[PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin LD_LIBRARY_PATH=/home/daniel/tmp/cuda_jetpack5:/tmp/ollama2043741824/runners/cuda_jetpack5:/tmp/ollama2043741824/runners CUDA_VISIBLE_DEVICES=GPU-de921cca-84a7-545a-ac50-34a5746dc088]"
...
ggml_cuda_init: found 1 CUDA devices:
  Device 0: Orin, compute capability 8.7, VMM: yes
llm_load_tensors: ggml ctx size =    0.27 MiB
llm_load_tensors: offloading 32 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 33/33 layers to GPU
...
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -18,17 +21,66 @@ ENV PATH /opt/rh/devtoolset-10/root/usr/bin:$PATH
...
+ARG CMAKE_VERSION
+RUN apt-get update && apt-get install -y git curl && \
+    curl -s -L https://github.com/Kitware/CMake/releases/download/v${CMAKE_VERSION}/cmake-${CMAKE_VERSION}-linux-$(uname -m).tar.gz | tar -zx -C /usr --strip-components 1
+COPY --from=llm-code / /go/src/github.com/ollama/ollama/
+WORKDIR /go/src/github.com/ollama/ollama/llm/generate
+ARG CGO_CFLAGS
+RUN OLLAMA_SKIP_STATIC_GENERATE=1 \
+    OLLAMA_SKIP_CPU_GENERATE=1 \
+    CUDA_VARIANT="_xavier" \
+    CMAKE_CUDA_ARCHITECTURES="80;86;87" \
```
Comment (Zunhammer) [inline Dockerfile:59]: I think the GPU architectures are wrong here. Regarding nvidia the architecture for Xavier is 72 (refer to https://developer.nvidia.com/cuda-gpus)

For Orin 87 without 80 and 87 should be fine I guess, and Jetson Nano is 53.

File: Dockerfile [hunk 2]
```diff
@@ -18,17 +21,66 @@ ENV PATH /opt/rh/devtoolset-10/root/usr/bin:$PATH
...
 COPY ./scripts/rh_linux_deps.sh /
 RUN CMAKE_VERSION=${CMAKE_VERSION} sh /rh_linux_deps.sh
 ENV PATH /opt/rh/gcc-toolset-10/root/usr/bin:$PATH
 COPY --from=llm-code / /go/src/github.com/ollama/ollama/
 WORKDIR /go/src/github.com/ollama/ollama/llm/generate
 ARG CGO_CFLAGS
-RUN OLLAMA_SKIP_STATIC_GENERATE=1 OLLAMA_SKIP_CPU_GENERATE=1 sh gen_linux.sh
+RUN OLLAMA_SKIP_STATIC_GENERATE=1 OLLAMA_SKIP_CPU_GENERATE=1 bash gen_linux.sh
+
+FROM --platform=linux/arm64 nvcr.io/nvidia/l4t-jetpack:${JETPACK_6} AS cuda-build-jetpack6-arm64
```
Comment (jmorganca) [inline Dockerfile:36]: Do you know if there's a way to make this opt-in by default?

Comment (dhiltgen) [reply Dockerfile:36]: Can you clarify?  Jetson users who build from source and link against the cuda libraries on their system work correctly.  The intent on this PR is to get our official binaries working on these platforms.

The current logic in this PR builds 3 variants for arm cuda
- Discrete GPUs (the original setup)
- JetPack 6
- JetPack 5

At runtime, we default to the discrete GPU version, but if we detect JP6 or JP5 environment variables we toggle to those respective iGPU runners.

**Comment by 631952531** [comment]
I wonder if this supports r36.3 version.

**Comment by dhiltgen** [comment]
> I wonder if this supports r36.3 version.

I don't have the ability to test this, but I would expect it should given the same major version.

**Comment by dhiltgen** [comment]
I found some corner cases while testing this on arm64 server systems which are fixed in #5049 which carries this set of changes for Jetson support, so I'll close this PR in favor of the other one.
